Fix incorrect header names in Webhook class

sphinks opened the pull request:
Fix minor problems in webhook class:
- correct name of custom 'help-scout' headers
- add mockito to test harness
- create small unit test for checking of headers

sphinks committed:
Fix minor problems in webhook class:
- correct name of custom 'help-scout' headers
- add mockito to test harness
- create small unit test for checking of event header
(cherry picked from commit 124d2ed)

sphinks committed:
- Improve tests for Webhook
- Add test json data for Webhook
(cherry picked from commit 7ddd96d)

sphinks committed:
- delete unsed extension of base Api Test Class

bob983 commented:
Hi @sphinks, thank you for the PR, I appreciate the fix and the tests(!). I'm going to check the PR tomorrow.

sphinks commented:
@bob983 that is ok. By the way there is the same problem in [developer's documentation](http://developer.helpscout.net/help-desk-api/webhooks/): PHP example contains wrong header name 'HTTP_X_HELPSCOUT_SIGNATURE'. Can you fix it?

Do you have any CI for building API lib?

sphinks committed:
- add integration test for check api for conversations

sphinks committed:
- add integration test for check api for users, tag, mailboxes
- extend test for conversations

bob983 commented:
Oh, sorry @sphinks, it's taking me ages! :( When I look into this, I'll check `HTTP_X_HELPSCOUT_SIGNATURE` as well.

As for CI - no there isn't one at the moment, but I recently added a lot of tests. We will start working on a new API version soon and it will have a proper CI.

sphinks commented:
@bob983 ok, I see, I just want to setup CI via codeship and some tests that I'm doing now. Do you have a place when I can put built jar?

P.S. Take a look at issues, that I have posted.

sphinks committed:
- create separate profile to run unit tests and integration test. Unit test are default profile

sphinks committed:
Merge pull request #1 from sphinks/feature/Integration_API_test

Feature/integration api test

sphinks commented:
@bob983 I have added some more commits occasionally to this pull-request. Actually I want to wait until you merger this pull request, but seem like all new commits are here too. What the are about:
1. Some kind of integration test that ping as much endpoint as it can to verify that API lib is not outdated. This test is suitable for CI run.
2. Added maven surefire plugin.
3. Created two profiles to run tests: by default it works as it was, using command 'mvn test -P=full-pack -DApiKey=<your_api_key>' it runs all tests including integration test.

Please, take look.

bob983 commented:
Oh nice, thanks for the other changes as well. I'm still busy with the API itself, but I did not forget about this SPR :)

sphinks commented:
@bob983 if you developing new API, may be you could give some document what will be added or changed, so I can create a new version of API wrapper in parallel?

P.S. Can I contact you directly somehow? I have some idea about improving helpscout with API, but current implementation not allowed to do it. I want to discuss it.

bob983 commented:
@sphinks It will take a while before the new API is ready (we haven't started) and we want to deliver a proper wrapper as well. You can contact me at bob[at]helpscout.com, we welcome feedback and ideas! :)

sphinks committed:
- fix NPE in case ErrorStream is null
Related issue: https://github.com/helpscout/helpscout-api-java/issues/23

bob983 commented:
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;ZSwf8ZfL&#x2F;149-take-a-look-at-public-pr-helpscout-api-java"><img src="https:&#x2F;&#x2F;github.trello.services&#x2F;images&#x2F;trello-icon.png" width="12" height="12"> Take a look at public PR helpscout-api-java</a>

sphinks commented:
@bob983 cannot get to card:

> Card not found.
> This card may be on a private board. If someone gave you this link, they may need to invite you to one of their boards or teams.

As far as I understand you have created trello card for yourself?:)

bob983 commented on src/test/java/net/helpscout/api/WebhookApiTest.java:
```diff
@@ -0,0 +1,109 @@
+     * x-forwarded-server: youcallback.domain.com
+     * connection: Keep-Alive
+     *
+     * Pay attention to case of header names - according to specification
```
What documentation do you have in mind? `getHeader()` is case insensitive : http://docs.oracle.com/javaee/1.3/api/javax/servlet/http/HttpServletRequest.html#getHeader%28java.lang.String%29. Or did you mean API documentation?

sphinks replied:
You are right - header name is case insensitive in java methods, but I was looking into HTTP specification (https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.2), but seem like I was reading it in the night, cause missed that header names are case insensitive. The comment looks irrelevant. Fix it and code.

bob983 commented:
@sphinks Yeah, that's card for me :smile:

bob983 commented on src/test/java/net/helpscout/api/WebhookApiTest.java:
```diff
@@ -0,0 +1,109 @@
+
+    @SneakyThrows
+    private static String readJsonDataToString(String jsonFileName) {
+        Path pathToJsonData = Paths.get(ClassLoader.getSystemResource("responses/" + jsonFileName + ".json").toURI());
```
The library is still for Java 1.6, sorry! :cry: `Path` is 1.7+

sphinks replied:
I get this with `Path` solution from AbstractApiClientTest (line 41), so seem like library was already moved to 1.7 before my commit :) I will search what can be used instead of Path in both places.

bob983 replied:
Ooops, I didn't know that :(

bob983 commented on src/test/java/net/helpscout/api/integration/ApiIntegrationTest.java:
```diff
@@ -0,0 +1,287 @@
+package net.helpscout.api.integration;
+
+import jdk.nashorn.internal.ir.annotations.Ignore;
```
I don't think this is the correct `@Ignore` annotation :)

sphinks replied:
Thanks. Fix it.

bob983 commented:
@sphinks I went through the PR, I like the unit tests for Webhooks. I left you one comment about using Java 1.7 code, because unfortunately, the library is still meant for Java 1.6+.

I'm still thinking about those integration tests. While they make sense for you and your test setup, I'm not yet sure they should be in this code base.

sphinks committed:
- delete irrelevant comment in tests about case problems of headers

sphinks committed:
- incorrect @Ignore annotation

sphinks commented:
@bob983, thanks for your review! I have fixed some problems, take a look, please.

Integration tests: not sure that they have sense only for me and my set up. Anyone who is going to use library can first of all run such test against the own account and ping available functions to be sure that library is not outdated. When I first time tried to use webhooks I faced the issue that library did not in actually works properly. But it is just one feature, main story about such tests - CI. Every commit from dev team ideally can go to CI engine and integration tests can be run against dev team test account, then CI moves ready jar to some repo for end users. It is not my personal tests, I just try to make the first step in creating CI process for the library. Of course, integration tests should be extended and CI process should be introduced, but make one step at one time :)

So those tests are surely related to this repository. What I agree with is the fact that new version of API and wrapper is going to development, so CI for old version won't be introduced. In this case, we can move this class to 'examples'.

bob983 commented:
#### Integration test

If the integration was run against development or sandboxed version of the API, I would be happy to keep it. But in current situation I don't think it belongs to the repository. I realise that not having a proper CI for the project is a bad thing, no doubt here and I also realise that you spent a lot of time creating that test and we all here appreciate it.

Could you please either remove the `pom.xml` change together with the test or come up with a way how to turn the test into example?

#### About Java 7:

If you haven't touched the `Path` yet, we might just change the description that the client now uses Java 7

#### Documentation

When this PR is merged, I'm going to fix the documentation as well.

sphinks committed:
- revert pom
- delete integration test (current version of wrapper will not support CI, so they are useless)

sphinks commented:
@bob983 I have deleted from pom special profiles for integration tests and delete integration tests (I did not convert them to an example, cause there is not so much useful code for usual working case).

I have left `Path` as it was.

P.S. You do not have to be sorry for asking me to delete those tests. :) Everything is ok, I suppose it is nothing, but just usual working process.:)

bob983 commented:
Thanks! Merging.
